File: sample-chat-obj-c/sample-chat/Components/ChatScreen/Views/ChatCollectionView/ChatCell/ChatCellLayoutAttributes.m
//
//  ChatCellLayoutAttributes.m
//  sample-chat
//
//  Created by Injoit on 2/25/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

#import "ChatCellLayoutAttributes.h"

@implementation ChatCellLayoutAttributes
//MARK: - Lifecycle

- (instancetype)init {
    
    self = [super init];
    if (self) {
        self.transform = CGAffineTransformMake(1, 0, 0, -1, 0, 0);
    }
    return self;
}

- (void)dealloc {
}

//MARK: - Setters

- (void)setContainerSize:(CGSize)containerSize {
    
    NSParameterAssert(containerSize.width >= 0.0f && containerSize.height >= 0.0f);
    _containerSize = [self correctedSizeFromSize:containerSize];
}

- (void)setAvatarSize:(CGSize)avatarSize {
    _avatarSize = [self correctedSizeFromSize:avatarSize];
}

//MARK: - Utilities

- (CGSize)correctedSizeFromSize:(CGSize)size {
    return CGSizeMake(ceilf(size.width), ceilf(size.height));
}

//MARK: - NSObject

- (BOOL)isEqual:(id)object {
    if (self == object) {
        return YES;
    }
    
    if (![object isKindOfClass:[self class]]) {
        
        return NO;
    }
    
    if (self.representedElementCategory == UICollectionElementCategoryCell) {
        
        ChatCellLayoutAttributes *layoutAttributes = (ChatCellLayoutAttributes *)object;
        
        if (!CGSizeEqualToSize(layoutAttributes.containerSize, self.containerSize)
            ||!CGSizeEqualToSize(layoutAttributes.avatarSize, self.avatarSize)
            ||!UIEdgeInsetsEqualToEdgeInsets(layoutAttributes.containerInsets, self.containerInsets)
            ||(int)layoutAttributes.topLabelHeight != (int)self.topLabelHeight
            ||(int)layoutAttributes.bottomLabelHeight != (int)self.bottomLabelHeight
            || (int)layoutAttributes.spaceBetweenTopLabelAndTextView != (int)self.spaceBetweenTopLabelAndTextView
            || (int)layoutAttributes.spaceBetweenTextViewAndBottomLabel != (int)self.spaceBetweenTextViewAndBottomLabel) {
            
            return NO;
        }
    }
    return [super isEqual:object];
}

- (NSString *)description {
    
    NSMutableString *desc = [NSMutableString stringWithString:[super description]];
    
    [desc appendFormat:@""];
    return desc.copy;
}

- (NSUInteger)hash {
    
    return [self.indexPath hash];
}

//MARK: - NSCopying

- (instancetype)copyWithZone:(NSZone *)zone {
    
    ChatCellLayoutAttributes *copy = [super copyWithZone:zone];
    
    if (copy.representedElementCategory != UICollectionElementCategoryCell) {
        return copy;
    }
    
    copy.avatarSize = self.avatarSize;
    copy.containerSize = self.containerSize;
    copy.containerInsets = self.containerInsets;
    copy.bottomLabelHeight = self.bottomLabelHeight;
    copy.topLabelHeight = self.topLabelHeight;
    copy.spaceBetweenTextViewAndBottomLabel = self.spaceBetweenTextViewAndBottomLabel;
    copy.spaceBetweenTopLabelAndTextView = self.spaceBetweenTopLabelAndTextView;
    
    return copy;
}

@end

File: sample-chat-swift/sample-chat-swift/Classes/Chat/CollectionView/ChatCellLayoutAttributes.swift
//
//  ChatCellLayoutAttributes.swift
//  Swift-ChatViewController
//
//  Created by Injoit on 1/28/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

import UIKit

struct ChatCellLayoutAttributesConstant {
    static let invalidParameter = "Invalid parameter not satisfying: containerSize.width >= 0.0 && containerSize.height >= 0.0"
}

class ChatCellLayoutAttributes: UICollectionViewLayoutAttributes {
    
    //MARK: - Properties
    private var _containerSize: CGSize = .zero
    var containerSize: CGSize {
        get {
            return _containerSize
        }
        set {
            var width = newValue.width
            var height = newValue.height
            width.round(.up)
            height.round(.up)
            let newSize = CGSize(width: width, height: height)
            _containerSize = newSize
        }
    }
    
    private var _avatarSize: CGSize = .zero
    var avatarSize: CGSize {
        get {
            return _avatarSize
        }
        set {
            var width = newValue.width
            var height = newValue.height
            width.round(.up)
            height.round(.up)
            let newSize = CGSize(width: width, height: height)
            _avatarSize = newSize
        }
    }
    
    
    var containerInsets = UIEdgeInsets.zero
    var topLabelHeight: CGFloat = 0.0
    var bottomLabelHeight: CGFloat = 0.0
    var spaceBetweenTopLabelAndTextView: CGFloat = 0.0
    var spaceBetweenTextViewAndBottomLabel: CGFloat = 0.0
    
    
    
    //MARK: - Lifecycle
    override init() {
        super.init()
        commonInit()
    }
    
    func commonInit() {
        self.transform = CGAffineTransform(a: 1, b: 0, c: 0, d: -1, tx: 0, ty: 0)
    }
    
    deinit {
    }
    
    //MARK: - Utilities
    
    override var hash: Int {
        return indexPath.hashValue
    }
    
    override func isEqual(_ object: Any?) -> Bool {
        guard let layoutAttributes = object as? ChatCellLayoutAttributes else {
            return false
        }
        
        if self === layoutAttributes {
            return true
        }
        
        if representedElementCategory == .cell {
            
            if !(layoutAttributes.containerSize.equalTo(containerSize)) ||
                !layoutAttributes.avatarSize.equalTo(avatarSize) ||
                !(layoutAttributes.containerInsets == containerInsets) ||
                Int(layoutAttributes.topLabelHeight) != Int(topLabelHeight) ||
                Int(layoutAttributes.bottomLabelHeight) != Int(bottomLabelHeight) ||
                Int(layoutAttributes.spaceBetweenTopLabelAndTextView) != Int(spaceBetweenTopLabelAndTextView) ||
                Int(layoutAttributes.spaceBetweenTextViewAndBottomLabel) != Int(spaceBetweenTextViewAndBottomLabel) {
                
                return false
            }
        }
        return super.isEqual(object)
    }
    
    
    override func copy(with zone: NSZone? = nil) -> Any {
        let copy = super.copy(with: zone)
        
        guard let cellLayout = copy as? ChatCellLayoutAttributes else {
            return copy
        }
        
        if cellLayout.representedElementCategory != .cell {
            return cellLayout
        }
        
        cellLayout.avatarSize = avatarSize
        cellLayout.containerSize = containerSize
        cellLayout.containerInsets = containerInsets
        cellLayout.bottomLabelHeight = bottomLabelHeight
        cellLayout.topLabelHeight = topLabelHeight
        cellLayout.spaceBetweenTextViewAndBottomLabel = spaceBetweenTextViewAndBottomLabel
        cellLayout.spaceBetweenTopLabelAndTextView = spaceBetweenTopLabelAndTextView
        
        return cellLayout
    }
    
}
